mesh/mesh.hpp

```diff
@@ -1272,46 +1275,47 @@ class Mesh
 
        The element indices in this structure don't need further processing,
        contrary to the ones obtained through Mesh::GetFacesElements and can
-       directly be used as Elem1 and Elem2 indices.
+       directly be used, e.g., Elem1 and Elem2 indices.
        Likewise the orientations for Elem1 and Elem2 already take into account
        special cases and can be used as is.
    */
    struct FaceInformation
    {
-      FaceLocation location; // Local, shared, boundary
-      FaceConformity conformity; // Conforming, non-conforming master or slave
+      FaceLocation elem_1_location, elem_2_location; // Local, shared, boundary
```

Comment:
The name `FaceLocation` is not good here -- these are properties of the two adjacent elements, so it is better to call the enum `ElementLocation`.

Also, "shared" location for the element is not a good term to use here -- in MFEM a geometric entity is called "shared" when it belongs to 2 or more ranks. The term we use to describe elements from other ranks that have a common face with an element on this rank is "face-neighbor".

Also, "boundary" location is unnecessary -- for boundary faces we can simply set `elem_1_location = Local` and `elem_2_location = NA`.

Reply:
Actually, `elem_1_location` is always going to be `Local`, and only `elem_2_location` can be `Local`, `FaceNeighbor`, or `NA`. So we can just store `elem_2_location`.

Reply:
I know, but it's the same with the orientation, a compiler is smart enough to optimize this, we don't need to do it ourselves.

Reply:
Besides, I don't think it is the same to assume that users will know the convention that `elem_1_location` is always `Local`, and to put this information in the code so that user don't have to know this convention.

Reply:
If `elem_1_location` is always guaranteed to be `Local`, you could do something like
```c++
struct FaceInformation
{
   static constexpr FaceLocation elem_1_location = FaceLocation::Local; // Element 1 is always guaranteed to be "Local"
   FaceLocation elem_2_location;
   […]
};
```
That way the invariant is stated explicitly rather than implicitly, and users could still access `elem_1_location` without knowing this detail and it should work correctly.

I think Yohann is right that this should not incur any real runtime or memory cost.

Reply:
I like @pazner's suggestion.

Reply:
I was about to do this change when I realized that it doesn't work with the current enum type, since `elem_1_location` can also be `Boundary`.

A solution that may make everyone happy may be to have:
```c++
enum class ElementLocation { Local, Shared, NA };
```
As @v-dobrev was mentioning we may be able to find a better name than `Shared`, as this describes more the face than a neighboring element. But `Neighbor` will not be very clear I think. `Ghost` could be confusing as "ghost faces" refers only to specific shared non-conforming faces (not all shared non-conforming faces are ghost faces).

And replace the two `FaceConformity` with a unique `FaceConfig` for the face:
```c++
enum class FaceConfig
{
   Boundary,
   Conforming,
   NonConformingFineCoarse, // elem1 is fine, elem2 is coarse
   NonConformingCoarseFine, // elem1 is coarse, elem2 is fine
   NA
};
```

I would put in `FaceConfig::NA` the "ghost master faces", and the "uninitialized ghost faces" because they are undistinguishable from each other and only contain `-1`s (unset values).
And I would also put the "local non-conforming master" faces in `FaceConfig::NA` since only "elem_1_index" is set (everything else is unset), making them mostly useless and only confusing I think. This would prevent users from having to systematically filter them out (and to have to understand that they don't contain usable information), for instance the code in #2195 would become more readable, as I could remove all the recurring:
```c++
      if ( face.IsLocalNonConformingCoarseFine() )
      {
         // We skip local non-conforming coarse-fine faces as they are treated
         // by the corresponding local non-conforming fine-coarse faces.
         continue;
      }
```
MFEM already systematically filters-out "local non-conforming master" faces.